starter-code/src/viking.js

```diff
@@ -1,11 +1,103 @@
 // Soldier
-function Soldier() {}
-
-// Viking
-function Viking() {}
-
-// Saxon
-function Saxon() {}
-
-// War
-function War() {}
+function Soldier (health, strength) {
+    this.health = health;
+    this.strength = strength;
+  };
+  
+  
+  Soldier.prototype.attack = function () {
+    return this.strength;
+  };
+  
+  Soldier.prototype.receiveDamage = function (damage) {
+    this.health =  this.health - damage;
+  };
+  
+  // Viking
+  function Viking(name, health, strength) {
+    Soldier.call(this, health, strength);
+    this.name = name;
+  };
+  
+  Viking.prototype = Object.create(Soldier.prototype);
+  Viking.prototype.constructor = Viking;
+  
+  Viking.prototype.receiveDamage = function(damage) {
+    this.health = this.health - damage;
+    if (this.health > 0) {
+      return this.name + ' has received ' + damage + ' points of damage';
+    } else {
+      return this.name + ' has died in act of combat';
+    }
+  };
+  
+  Viking.prototype.battleCry = function () {
+    return "Odin Owns You All!";
+  };
+  
+  // Saxon
+  function Saxon(health, strength) {
+  Soldier.call(this, health, strength);
+  }
+  Saxon.prototype = Object.create(Soldier.prototype);
+  Saxon.prototype.constructor = Saxon;
+  
+  Saxon.prototype.receiveDamage = function(damage) {
+    this.health = this.health - damage;
+    if (this.health > 0) {
+      return 'A Saxon has received ' + damage + ' points of damage';
+    } else {
+      return 'A Saxon has died in combat';
+    }
+  };
+  
+  
+  // War
+  function War() {
+    this.vikingArmy = [];
+    this.saxonArmy = [];
+  };
+  
+  War.prototype.addViking = function(viking) {
+    this.vikingArmy.push(viking);
+  };
+  
+  War.prototype.addSaxon = function(saxon){
+    this.saxonArmy.push(saxon);
+  };
+  
+  War.prototype.vikingAttack = function() {
+    var randomVicking = this.vikingArmy[Math.floor(Math.random() * this.vikingArmy.length)]
+    var randomSaxon = this.saxonArmy[Math.floor(Math.random() * this.saxonArmy.length)]
+  
+    var valueToReturnSaxon = randomSaxon.receiveDamage(randomVicking.strength);
+  
+    this.saxonArmy = this.saxonArmy.filter(function(saxon) {
+      return saxon.health > 0;
+    });
```

Comment:
No está mal pensado, pero estáis recorriendo el array siempre para devolver otro array con todos los elementos que quedan, incluso si no ha muerto nadie. Si solo evaluáis el elemento que ha recibido el daño sería mejor